wgpu-hal/src/dx12/suballocation.rs

```diff
@@ -0,0 +1,307 @@
+};
+
+// This exists to work around https://github.com/gfx-rs/wgpu/issues/3207
+// Currently this will work the older, slower way if the windows_rs feature is disabled,
+// and will suballocate buffers using gpu_allocator if the windows_rs feature is enabled.
+
+#[cfg(feature = "windows_rs")]
+use gpu_allocator::{
+    d3d12::{AllocationCreateDesc, ToWinapi, ToWindows},
+    MemoryLocation,
+};
+
+// TODO: find the exact value
+// TODO: figure out if this is even needed?
+#[allow(unused)] // TODO: Exists until windows-rs is standard, then it can probably be removed?
+pub(crate) const D3D12_HEAP_FLAG_CREATE_NOT_ZEROED: u32 = d3d12::D3D12_HEAP_FLAG_NONE;
+
+#[derive(Debug)]
+pub(crate) struct GpuAllocatorWrapper {
+    #[cfg(feature = "windows_rs")]
+    pub(crate) allocator: gpu_allocator::d3d12::Allocator,
+    #[cfg(not(feature = "windows_rs"))]
+    #[allow(unused)]
+    pub(crate) allocator: (),
+}
+
+#[derive(Debug)]
+pub(crate) struct AllocationWrapper {
+    #[cfg(feature = "windows_rs")]
+    pub(crate) allocation: gpu_allocator::d3d12::Allocation,
+    #[cfg(not(feature = "windows_rs"))]
+    #[allow(unused)]
+    pub(crate) allocation: (),
+}
+
+#[cfg(feature = "windows_rs")]
+pub(crate) fn create_allocator_wrapper(
+    raw: &native::Device,
+) -> Result<Option<Mutex<GpuAllocatorWrapper>>, crate::DeviceError> {
+    use log::error;
+
+    let device = raw.as_ptr();
+
+    match gpu_allocator::d3d12::Allocator::new(&gpu_allocator::d3d12::AllocatorCreateDesc {
+        device: device.as_windows().clone(),
+        debug_settings: Default::default(),
+    }) {
+        Ok(allocator) => Ok(Some(Mutex::new(GpuAllocatorWrapper { allocator }))),
+        Err(e) => {
+            error!("Failed to create d3d12 allocator, error: {}", e);
+            Err(e)?
+        }
+    }
+}
+
+#[cfg(not(feature = "windows_rs"))]
+pub(crate) fn create_allocator_wrapper(
+    _raw: &native::Device,
+) -> Result<Option<Mutex<GpuAllocatorWrapper>>, crate::DeviceError> {
+    Ok(None)
+}
+
+#[cfg(feature = "windows_rs")]
+pub(crate) fn create_buffer_resource(
+    device: &super::Device,
+    desc: &crate::BufferDescriptor,
+    raw_desc: d3d12::D3D12_RESOURCE_DESC,
+    resource: &mut WeakPtr<ID3D12Resource>,
+) -> Result<(HRESULT, Option<AllocationWrapper>), crate::DeviceError> {
+    let is_cpu_read = desc.usage.contains(crate::BufferUses::MAP_READ);
+    let is_cpu_write = desc.usage.contains(crate::BufferUses::MAP_WRITE);
+    // TODO: These are probably wrong?
+    let location = match (is_cpu_read, is_cpu_write) {
+        (true, true) => MemoryLocation::CpuToGpu,
+        (true, false) => MemoryLocation::GpuToCpu,
+        (false, true) => MemoryLocation::CpuToGpu,
+        (false, false) => MemoryLocation::GpuOnly,
+    };
+
+    let name = desc.label.unwrap_or("Unlabeled buffer");
+
+    // SAFETY: allocator exists when the windows_rs feature is enabled
+    let mut allocator = unsafe { device.mem_allocator.as_ref().unwrap_unchecked().lock() };
```

Comment:
Now that #3051 is in, let's make an extension trait in assertions.rs like:

```rust
trait StrictAssertUnwrapExt<T> {
    fn strict_unwrap_unchecked(self) -> T;
}

impl<T> StrictAssertUnwrapExt<T> for Option<T> {
    fn strict_unwrap_unchecked(self) -> T {
        strict_assert!(self.is_some(), "Called strict_unwrap_unchecked on None");
        unsafe { self.unwrap_unchecked() }
    }
}

impl<T, E> StrictAssertUnwrapExt<T> for Result<T, E> {
    fn strict_unwrap_unchecked(self) -> T {
        strict_assert!(self.is_ok(), "Called strict_unwrap_unchecked on Err");
        unsafe { self.unwrap_unchecked() }
    }
}
```

Reply:
I wasn't sure if you meant the assertions.rs in `wgpu-types`, or the one in `wgpu-core`, but to avoid a circular dependency I implemented this in the wgpu-types one and changed its `mod assert` to public.

Reply:
Ack, the one in core should have gotten removed. Types was the right decision - I'll sort the assertions out.